eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/DualFundingHandlers.scala

```diff
@@ -28,26 +34,79 @@ import scala.util.{Failure, Success}
 /**
  * This trait contains handlers related to dual-funding channel transactions.
  */
-trait DualFundingHandlers extends CommonHandlers {
+trait DualFundingHandlers extends CommonFundingHandlers {
 
   this: Channel =>
 
-  def publishFundingTx(d: DATA_WAIT_FOR_DUAL_FUNDING_PLACEHOLDER): Unit = {
-    d.fundingTx match {
+  def publishFundingTx(fundingParams: InteractiveTxParams, fundingTx: SignedSharedTransaction): Unit = {
+    fundingTx match {
       case _: PartiallySignedSharedTransaction =>
         log.info("we haven't received remote funding signatures yet: we cannot publish the funding transaction but our peer should publish it")
       case fundingTx: FullySignedSharedTransaction =>
         // Note that we don't use wallet.commit because we don't want to rollback on failure, since our peer may be able
         // to publish and we may be able to RBF.
         wallet.publishTransaction(fundingTx.signedTx).onComplete {
           case Success(_) =>
-            context.system.eventStream.publish(TransactionPublished(d.commitments.channelId, remoteNodeId, fundingTx.signedTx, fundingTx.tx.localFees(d.fundingParams), "funding"))
-            channelOpenReplyToUser(Right(ChannelOpenResponse.ChannelOpened(d.commitments.channelId)))
+            context.system.eventStream.publish(TransactionPublished(fundingParams.channelId, remoteNodeId, fundingTx.signedTx, fundingTx.tx.localFees(fundingParams), "funding"))
+            channelOpenReplyToUser(Right(ChannelOpenResponse.ChannelOpened(fundingParams.channelId)))
           case Failure(t) =>
             channelOpenReplyToUser(Left(LocalError(t)))
             log.warning("error while publishing funding tx: {}", t.getMessage) // tx may be published by our peer, we can't fail-fast
         }
     }
   }
 
+  def handleDualFundingConfirmedOffline(w: WatchFundingConfirmedTriggered, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED) = {
+    if (w.tx.txid == d.commitments.commitInput.outPoint.txid) {
+      watchFundingTx(d.commitments)
+      context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
```

Comment:
Sending `TransactionConfirmed` here will lead to a duplicate (we will send them again in state `WAIT_FOR_DUAL_FUNDING_CONFIRMED` after we reconnect), and is different from the single-funding flow (where we won't emit that event if the funding confirms when disconnected and then we go directly to close).

It troubles me a little bit that this handler is very similar, but not quite the same as in `WAIT_FOR_DUAL_FUNDING_CONFIRMED`. For example, we don't call `Transaction.correctlySpends`.

Reply:
Now that I think about it: why do we call `Transaction.correctlySpends` at all in `Channel.scala`? In the dual funding case we already know the full funding tx and have already verified it in `InteractiveTxBuilder.scala`.

Reply:
> Sending TransactionConfirmed here will lead to a duplicate

This isn't an issue, this is deduplicated in the DB. I believe it's important to send it here, because if we close before reconnecting, we would otherwise never send it (which is probably a bug in the single-funding flow, that we could fix when updating the single-funding flow to match the dual-funded one and rely on watching confirmation first?).

> For example, we don't call Transaction.correctlySpends.

We could, calling this function is really just a belts-and-suspenders because of the first CVE lightning had, it should never yield a failure, but I agree that it would be more symmetrical.

Reply:
> Now that I think about it: why do we call Transaction.correctlySpends at all in Channel.scala? In the dual funding case we already know the full funding tx and have already verified it in InteractiveTxBuilder.scala.

Let's just remove it, it simplifies the code!

Reply:
Removed in https://github.com/ACINQ/eclair/pull/2274/commits/f49116ab7af3733bd2b1055c66e8eaa27236276b

Reply:
I think we should rollback alternative funding txs before "forgetting" them. Actually I believe there is a bug in the `WAIT_FOR_DUAL_FUNDING_CONFIRMED` handler: if the latest funding tx confirms, it automatically rollbacks all alternative txs (since they are double spent), but the opposite isn't true if an alternative tx confirms.

I also think that we should we rollback other funding txs as soon as possible, i.e. in `OFFLINE`/`SYNCING`. We may stay disconnected for a long time.

Nit: There are minor inconsistencies in logs between the two handlers.

Reply:
I harmonized the logs in https://github.com/ACINQ/eclair/pull/2274/commits/ccf43f9202609d78d053d36f1e27949365461944

Note that in the offline/syncing and closing cases, we don't log anything when the expected funding transaction confirms, because this logs would otherwise repeat every time we restart eclair, which is unnecessarily noisy. The node operator will get a log line when the channel comes back online or if it's closed before that, I believe this is enough?